agent/xds/xdscommon/xdscommon.go

```diff
@@ -97,70 +127,91 @@ func MakePluginConfiguration(cfgSnap *proxycfg.ConfigSnapshot) PluginConfigurati
-
+		// These are the discovery chains for upstreams which have the Envoy Extensions applied to the local service.
 		for uid, dc := range cfgSnap.ConnectProxy.DiscoveryChain {
-			if _, ok := connectProxies[uid]; !ok {
-				continue
-			}
-
-			serviceConfigs[upstreamIDToCompoundServiceName(uid)] = ServiceConfig{
-				Kind:            api.ServiceKindConnectProxy,
-				EnvoyExtensions: convertEnvoyExtensions(dc.EnvoyExtensions),
-			}
-
 			compoundServiceName := upstreamIDToCompoundServiceName(uid)
+			extensionConfigs[compoundServiceName] = convertEnvoyExtensions(dc.EnvoyExtensions)
+
 			meta := uid.EnterpriseMeta
 			sni := connect.ServiceSNI(uid.Name, "", meta.NamespaceOrDefault(), meta.PartitionOrDefault(), cfgSnap.Datacenter, trustDomain)
-			sniMappings[sni] = compoundServiceName
-			envoyIDMappings[uid.EnvoyID()] = compoundServiceName
+			outgoingKind, ok := outgoingKindByService[compoundServiceName]
+			if !ok {
+				outgoingKind = api.ServiceKindConnectProxy
+			}
+
+			upstreamMappings[compoundServiceName] = UpstreamData{
+				SNI:               map[string]struct{}{sni: {}},
+				EnvoyID:           uid.EnvoyID(),
+				OutgoingProxyKind: outgoingKind,
+			}
 		}
 	case structs.ServiceKindTerminatingGateway:
+		kind = api.ServiceKindTerminatingGateway
 		for svc, c := range cfgSnap.TerminatingGateway.ServiceConfigs {
 			compoundServiceName := serviceNameToCompoundServiceName(svc)
-			serviceConfigs[compoundServiceName] = ServiceConfig{
-				EnvoyExtensions: convertEnvoyExtensions(c.EnvoyExtensions),
-				Kind:            api.ServiceKindTerminatingGateway,
-			}
+			extensionConfigs[compoundServiceName] = convertEnvoyExtensions(c.EnvoyExtensions)
 
 			sni := connect.ServiceSNI(svc.Name, "", svc.NamespaceOrDefault(), svc.PartitionOrDefault(), cfgSnap.Datacenter, trustDomain)
-			sniMappings[sni] = compoundServiceName
-
 			envoyID := proxycfg.NewUpstreamIDFromServiceName(svc)
-			envoyIDMappings[envoyID.EnvoyID()] = compoundServiceName
+
+			snis := map[string]struct{}{sni: {}}
 
 			resolver, hasResolver := cfgSnap.TerminatingGateway.ServiceResolvers[svc]
 			if hasResolver {
 				for subsetName := range resolver.Subsets {
 					sni := connect.ServiceSNI(svc.Name, subsetName, svc.NamespaceOrDefault(), svc.PartitionOrDefault(), cfgSnap.Datacenter, trustDomain)
-					sniMappings[sni] = compoundServiceName
+					snis[sni] = struct{}{}
 				}
 			}
+
+			upstreamMappings[compoundServiceName] = UpstreamData{
+				SNI:               snis,
+				EnvoyID:           envoyID.EnvoyID(),
+				OutgoingProxyKind: api.ServiceKindTerminatingGateway,
+			}
+
 		}
 	}
 
-	return PluginConfiguration{
-		ServiceConfigs:       serviceConfigs,
-		SNIToServiceName:     sniMappings,
-		EnvoyIDToServiceName: envoyIDMappings,
-		Kind:                 api.ServiceKind(cfgSnap.Kind),
+	extensionConfigurations := make(map[api.CompoundServiceName][]ExtensionConfiguration)
+	for svc, exts := range extensionConfigs {
```

Comment:
I think it might help out with readability to change the name of either `extensionConfigurations` or `extensionConfigs`.

Reply:
Yeah good call, renamed to extensionConfigurationsMap and extensionsMap